src/graph.ts

```diff
@@ -1,30 +1,88 @@
 
-const graphCache = new Map<PackageInfos, [string, string][]>();
+import { createDependencyMap } from "./createDependencyMap";
+import multimatch from "multimatch";
 
-export function getPackageGraph(packages: PackageInfos) {
-  if (graphCache.has(packages)) {
-    return graphCache.get(packages)!;
-  }
+// Reference: https://github.com/pnpm/pnpm/blob/597047fc056dd25b83638a9ab3df0df1c555ee49/packages/filter-workspace-packages/src/parsePackageSelector.ts
+export interface PackageGraphScope {
+  namePatterns?: string[];
+  includeDependencies?: boolean;
+  includeDependents?: boolean;
+}
+
+export function createPackageGraph(packages: PackageInfos, scope: PackageGraphScope = {}): PackageGraph {
+  const dependencyMap = createDependencyMap(packages);
 
-  const edges: [string, string][] = [];
+  const packageSet = new Set<string>();
+  const edges: PackageGraph["dependencies"] = [];
 
-  for (const [pkg, info] of Object.entries(packages)) {
-    const deps = getInternalDeps(info, packages);
-    for (const dep of deps) {
-      edges.push([dep, pkg]);
+  const visitor: PackageGraphVisitor = (pkg, dependencies, dependents) => {
+    packageSet.add(pkg);
+    
+    if (scope.includeDependencies && dependencies) {
+      for (const dep of dependencies) {
+        edges.push({ name: pkg, dependency: dep });
+        packageSet.add(dep);
+      }
     }
-  }
 
-  graphCache.set(packages, edges);
+    if (scope.includeDependents && dependents) {
+      for (const dep of dependents) {
+        edges.push({ name: dep, dependency: pkg });
+        packageSet.add(dep);
+      }
+    }
+  };
+
+  visitPackageGraph(packages, dependencyMap, visitor, scope);
 
-  return edges;
+  return {packages: [...packageSet], dependencies: edges};
 }
 
-/**
- * @internal resets the graph cache for internal testing purpose only
- */
-export function _resetGraphCache() {
-  graphCache.clear();
+
+function visitPackageGraph(
+  packages: PackageInfos,
+  dependencyMap: DependencyMap,
+  visitor: PackageGraphVisitor,
+  scope: PackageGraphScope
+) {
+  const visited = new Set<string>();
+  const packageNames = Object.keys(packages);
+
+  const stack: string[] =
+    scope && scope.namePatterns ? multimatch(packageNames, scope.namePatterns) : packageNames;
+
+  while (stack.length > 0) {
+    const pkg = stack.pop()!;
+
+    if (visited.has(pkg)) {
+      continue;
+    }
+
+    const nextPkgs: string[] = [];
+    let dependencies: string[] = [];
+    let dependents: string[] = [];
+
+    if (scope?.includeDependencies) {
+      dependencies = [...dependencyMap.dependencies.get(pkg) ?? []];
+      nextPkgs.push(...dependencies);
+    }
+
+    if (scope?.includeDependents) {
+      dependents = [...dependencyMap.dependents.get(pkg) ??[]];
+      nextPkgs.push(...dependents);
+    }
+
+    visitor(pkg, dependencies, dependents);
+
+    visited.add(pkg);
+
+    if (nextPkgs.length > 0) {
+      for (const nextPkg of nextPkgs) {
+        stack.push(nextPkg);
```

Comment:
Packages for visiting is expanded by the discovered packages (based on the flags)